pkg/patterns/declarative/metrics.go

```diff
@@ -0,0 +1,342 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package declarative
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative/pkg/manifest"
+)
+
+const DefaultNamespace = "default"
+const ClusterScoped = "_Cluster_"
+
+const (
+	Declarative = "declarative_reconciler"
+)
+
+const (
+	ReconcileCount   = "reconcile_count"
+	ReconcileFailure = "reconcile_failure_count"
+
+	ManagedObjectsRecord = "managed_objects_record"
+)
+
+var (
+	reconcileCount = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: Declarative,
+		Name:      ReconcileCount,
+		Help:      "How many times reconciliation of K8s objects managed by declarative reconciler occurs",
+	}, []string{"group_version_kind", "namespace", "name"})
+
+	reconcileFailure = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: Declarative,
+		Name:      ReconcileFailure,
+		Help:      "How many times reconciliation failure of K8s objects managed by declarative reconciler occurs",
+	}, []string{"group_version_kind", "namespace", "name"})
+
+	managedObjectsRecord = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Subsystem: Declarative,
+		Name:      ManagedObjectsRecord,
+		Help:      "Track the number of objects in manifest",
+	}, []string{"group_version_kind", "namespace", "name"})
+)
+
+func gvkString(gvk schema.GroupVersionKind) string {
+	if len(gvk.Group) == 0 && gvk.Version == "v1" {
+		return gvk.Version + "/" + gvk.Kind
+	}
+
+	return gvk.Group + "/" + gvk.Version + "/" + gvk.Kind
+}
+
+func namespaced(mgr manager.Manager, gvk schema.GroupVersionKind) (bool, error) {
+	mapping, err := mgr.GetRESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return false, err
+	}
+
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+type reconcileCounter struct {
```

Comment:
One thing you could here is to have something like `reconcileMetrics` and expose a `ReconcileSuceeded` and `ReconcileFailed` method, and not expose the metric structure externally.  Then e.g. you could have `ReconcileFailed` take an error, if in future you wanted to handle some errors differently (a different metric) you could do that entirely in this metrics layer.

Reply:
I'll create new type `reconcileMetrics` in `metrics.go` file and use it. That type will implement similar method you suggest.